### PR TITLE
Nested test spans

### PR DESCRIPTION
**Pull request by felixscheinost**
PR for the improvement discussed in #146

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -144,6 +144,7 @@ openTelemetryBuild {
 | customTags              | `Map<String, String>`       | `null`                           | Custom tags to add to each trace                                                                                             |
 | parentSpanIdEnvVarName  | `String`                    | `SPAN_ID`                        | Environment variable name for a remote parent span ID (if using a parent trace like the Jenkins OpenTelemetry plugin)        |
 | parentTraceIdEnvVarName | `String`                    | `TRACE_ID`                       | Environment variable name for a remote parent trace ID (if using a parent trace like the Jenkins OpenTelemetry plugin)       |
+| hierarchicalTestSpans   | `Boolean`                   | `true`                           | Whether test spans should be hierarchical (per test executor, per test class, per method) or flat (just per test method)     |
```
Comment (felixscheinost) [inline README.md:147]: I don't really like the name of this option. Do you have an idea for a more fitting one?

File: src/main/kotlin/com/atkinsondev/opentelemetry/build/OpenTelemetryBuildPlugin.kt [hunk 2]
```diff
@@ -89,7 +89,14 @@ class OpenTelemetryBuildPlugin : Plugin<Project> {
...
 
-                        val taskListener = OpenTelemetryTaskListener(tracer, rootSpan, baggage, project.logger)
+                        val taskListener =
+                            OpenTelemetryTaskListener(
+                                tracer = tracer,
+                                rootSpan = rootSpan,
+                                baggage = baggage,
+                                logger = project.logger,
+                                hierarchicalTestSpans = extension.hierarchicalTestSpans.get(),
+                            )
```
Comment (felixscheinost) [inline src/main/kotlin/com/atkinsondev/opentelemetry/build/OpenTelemetryBuildPlugin.kt:99]: At first I wanted to pass in the whole `extension` into the listener. 

(Even though I already thought that one could argue that this would be a unnecessary tight coupling between the two classes).

I went with just `hierarchicalTestSpans` because of the tests as otherwise I'd have had to mock the extension somehow.

File: src/main/kotlin/com/atkinsondev/opentelemetry/build/OpenTelemetryTestListener.kt [hunk 3]
```diff
@@ -78,20 +77,44 @@ class OpenTelemetryTestListener(
             testTaskSpan.addEvent(TEST_FAILURE_SPAN_EVENT_NAME_KEY, attributesBuilder.build())
         }
 
-        val span = testSpanMap[testKey]
+        val span = testSpanMap.remove(test) ?: return
```
Comment (felixscheinost) [inline src/main/kotlin/com/atkinsondev/opentelemetry/build/OpenTelemetryTestListener.kt:80]: I personally like early returns but I agree that it *does* make control flow slightly less obvious. Should I undo the change here?

File: README.md (hunk 1 above)
Comment (craigatk) [reply README.md:147]: Hmm, maybe `nestedTestSpans` ?

File: src/main/kotlin/com/atkinsondev/opentelemetry/build/OpenTelemetryBuildPlugin.kt (hunk 2 above)
Comment (craigatk) [reply src/main/kotlin/com/atkinsondev/opentelemetry/build/OpenTelemetryBuildPlugin.kt:99]: +1 to only passing in the test-spans flag to avoid tightly coupling the extension class to the listener, as you mentioned.

File: src/main/kotlin/com/atkinsondev/opentelemetry/build/OpenTelemetryTestListener.kt (hunk 3 above)
Comment (craigatk) [reply src/main/kotlin/com/atkinsondev/opentelemetry/build/OpenTelemetryTestListener.kt:80]: Good question, I'm also all for early returns - but the `return` on the right side of an `?:` does read a little odd to me.

What about making the null check a bit more obvious / verbose? Something like:

```
if (span == null) {
  return
}
```

Comment (felixscheinost) [reply src/main/kotlin/com/atkinsondev/opentelemetry/build/OpenTelemetryTestListener.kt:80]: At least Jetbrains seems to consider `?: return` more idiomatic as IDEA sugguests using elvis when doing a `if (... == null) { return }`.

I'll just leave it as it was before 😃
